includes/handler.inc

```diff
@@ -430,46 +473,184 @@ function islandora_oai_query_solr($start = 0, $field = 'PID', $query = NULL, $se
+  $qp->solrParams['fq'] = array();
+  if (module_exists('islandora_xacml_api')) {
+    module_invoke('islandora_xacml_api', 'islandora_solr_query', $qp);
+  }
+  // Is part of the current collection.
+  $collection_fields = explode("\n", variable_get('islandora_oai_collection_field', "RELS_EXT_isMemberOfCollection_uri_ms\nRELS_EXT_isMemberOf_uri_ms"));
+  $collection_filter = array();
+  foreach ($collection_fields as $collection_field) {
+    $collection_filter[] = format_string('!field:("!set" OR "info:fedora/!set")', array(
+      '!field' => $collection_field,
+      '!set' => $parent,
+    ));
+  }
+  $qp->solrParams['fq'][] = '(' . implode(' OR ', $collection_filter) . ')';
+  // Is also a collection itself.
+  $qp->solrParams['fq'][] = format_string('!field:"info:fedora/islandora:collectionCModel"', array(
+    '!field' => variable_get('islandora_oai_content_model_field', 'RELS_EXT_hasModel_uri_ms'),
+  ));
 
-      $connection = islandora_get_tuque_connection();
-      if (!variable_get('islandora_risearch_use_itql_when_necessary', TRUE)) {
-        $role_filters = array();
-        foreach ($user->roles as $role) {
-          $role_filters[] = "?role = \"$role\"";
-        }
-        $xacml_filters = array(
-          '!bound(?role) && !bound(?user)',
-          'bound(?role) && (' . implode(' || ', $role_filters) . ')',
-          "bound(?user) && ?user = '$user_name'",
-        );
-        $parenthesis_wrap = function($string) {
-          return "($string)";
-        };
-        $xacml_filter = implode(' || ', array_map($parenthesis_wrap, $xacml_filters));
-        $query = <<<EOQ
+  $qp->executeQuery(FALSE);
+  $get_pid = function($object) {
+    return $object['PID'];
+  };
+  $pids = array_map($get_pid, $qp->islandoraSolrResult['response']['objects']);
+  $diff = array_diff($pids, $sets);
+  $sets = array_merge($sets, $pids);
+  foreach ($diff as $child_collection) {
+    islandora_oai_query_solr_for_child_sets($child_collection, $sets);
+  }
+  return $sets;
+}
+
+/**
+ * Helper function to query for child sets from the RI.
+ *
+ * @param string $set
+ *   The PID of the set to query in.
+ *
+ * @return array
+ *   The results of the RI query from Tuque.
+ */
+function islandora_oai_query_ri_for_child_sets($set) {
+  global $user;
+  $user_name = $user->uid === 0 ? 'anonymous' : $user->name;
+
+  $connection = islandora_get_tuque_connection();
+  // Query sets from RI (SPARQL).
+  if (!variable_get('islandora_risearch_use_itql_when_necessary', TRUE)) {
+    $role_filters = array();
+    foreach ($user->roles as $role) {
+      $role_filters[] = "?role = \"$role\"";
+    }
+    $xacml_filters = array(
+      '!bound(?role) && !bound(?user)',
+      'bound(?role) && (' . implode(' || ', $role_filters) . ')',
+      "bound(?user) && ?user = '$user_name'",
+    );
+    $parenthesis_wrap = function($string) {
+      return "($string)";
+    };
+    $xacml_filter = implode(' || ', array_map($parenthesis_wrap, $xacml_filters));
```

Comment:
I realize [it did it previously](https://github.com/Islandora/islandora_oai/pull/84/files#diff-d6fe00e82b5d432596ebff868818181bL471), but could we clean this up by [moving to](https://github.com/Islandora/islandora_xacml_editor/blob/7.x/api/islandora_xacml_api.module#L204-L223):
```php
module_invoke('islandora_xacml_api', 'islandora_basic_collection_get_query_filters');
```
... Appears that it shouldn't even be necessary to have a `module_exists()` (the return would just be `NULL`).